XamlAnimatedGif/Animator.cs

```diff
@@ -49,40 +49,41 @@ internal Animator(Stream sourceStream, Uri sourceUri, GifDataStream metadata, Re
 
             _cacheFrameDataInMemory = cacheFrameDataInMemory;
 
-            if (!_cacheFrameDataInMemory)
-                return;
+            if (cacheFrameDataInMemory)
+            {
+                _cachedFrameBytes = new byte[_metadata.Frames.Count][][];
+                _loadFramesDataTask = Task.Run(LoadFrames, CancellationToken.None);
+            }
+        }
 
-            var frameDescs = new GifImageDescriptor[_metadata.Frames.Count];
-            var frameRows = new IEnumerable<int>[_metadata.Frames.Count];
+        private void LoadFrames()
+        {
+            var biggestFrameSize = 0L;
             for (var frameIndex = 0; frameIndex < _metadata.Frames.Count; frameIndex++)
             {
-                var currentFrame = _metadata.Frames[frameIndex];
-                var frameDesc = currentFrame.Descriptor;
-                var currentRect = GetFixedUpFrameRect(frameDesc);
-                var currentFrameRows = frameDesc.Interlace
-                    ? InterlacedRows(currentRect.Height)
-                    : NormalRows(currentRect.Height);
-                
-                frameDescs[frameIndex] = frameDesc;
-                frameRows[frameIndex] = currentFrameRows.ToArray();
+                var startPosition = _metadata.Frames[frameIndex].ImageData.CompressedDataStartOffset;
+                var endPosition = _metadata.Frames.Count == frameIndex + 1
+                    ? _sourceStream.Length
+                    : _metadata.Frames[frameIndex + 1].ImageData.CompressedDataStartOffset - 1;
+                var size = endPosition - startPosition;
+                biggestFrameSize = Math.Max(size, biggestFrameSize);
             }
-            _loadFramesDataTask = Task.Run(() =>
+
+            byte[] indexCompressedBytes = new byte[biggestFrameSize];
+            for (var frameIndex = 0; frameIndex < _metadata.Frames.Count; frameIndex++)
             {
-                for (var frameIndex = 0; frameIndex < _metadata.Frames.Count; frameIndex++)
+                var frame = _metadata.Frames[frameIndex];
+                var frameDesc = _metadata.Frames[frameIndex].Descriptor;
+                GetIndexBytesAsync(frameIndex, indexCompressedBytes, CancellationToken.None).Wait();
```

Comment:
Don't use `.Wait()`; `await` the call instead, and change the `LoadFrames` method to `async Task`.

Reply:
good idea, I changed LoadFrames to async method